### PR TITLE
new function of clipping when generating L2 corrections, better fit used plotting L1 offsets.

### PR DESCRIPTION
**Pull request by sifuluo**

### CODE REVIEW

File: JetAnalyzers/test/run_JRA_cfg.py [hunk 1]
```diff
@@ -76,7 +76,11 @@
 except ImportError:
     print "Couldn't open the external list of files from DAS. If you just checkout out the JetResponseAnalyzer package you will need to make this file yourself. Currently Falling back to opening the list hard-coded in run_JRA_cfg.py. This is not a bad action as long as it is what you intended to have happen."
     inputFiles = cms.untracked.vstring(
-	    'root://cmsxrootd.fnal.gov//store/mc/<path to root file>/<filename>.root',
+	    # 'root://cmsxrootd.fnal.gov//store/mc/<path to root file>/<filename>.root'
+        # 'root://cmsxrootd.fnal.gov//store/mc/RunIIFall17DRPremix/QCD_Pt-15to7000_TuneCP5_Flat_13TeV_pythia8/AODSIM/94X_mc2017_realistic_v10-v1/50000/00304636-1BDB-E711-B6F3-FA163ECE02A9.root',
+        # 'root://cmsxrootd.fnal.gov//store/mc/RunIIFall17DRPremix/QCD_Pt-15to7000_TuneCP5_Flat_13TeV_pythia8/AODSIM/94X_mc2017_realistic_v10-v1/50000/0036C92E-DFDB-E711-952A-008CFAFC05DE.root',
+		'root://cms-xrd-global.cern.ch//store/user/kirschen/QCD_Pt-15to7000_TuneCP5_Flat2017_13TeV_pythia8/crab_pickEvents/180201_223255/0000/pickevents_1.root'
+		# /afs/cern.ch/user/k/kirschen/public/forJERC/forMCTruthDebugging/pickevents_NoPU.root
```
Comment (aperloff) [inline JetAnalyzers/test/run_JRA_cfg.py]: Please remove these references to a specific file. Keep only the example file path on line 79.

**Comment by sifuluo** [comment]
I have already deleted that part. I am not sure if the change has taken place.

**Comment by karavdin** [comment]
Hi @sifuluo @aperloff,
Are there any problems with this PR, so it can't be merged? @raggleton might profit of it, he asked to do the same clipping. 
@sifuluo, it is possible to make clipping flexible, i.e 8 GeV for AK4 and 17 GeV for AK8, please? At the moment we have to "fix" AK8 txt files manually.

Cheers,
Anastasia

**Comment by sifuluo** [comment]
The clipping is optional and set to 0 in L2 creator, In Closure maker it is not flexible yet but can be easily fixed.

**Comment by sifuluo** [comment]
The Clipping is flexible in L2 Creator. And by default, it is set to 0.0 . And In my iteration it was set to 8.0 .

**Comment by FHead** [comment]
Hi @sifuluo @aperloff,

What is the plan for this PR?  It would be nice to have all developments in so everyone can benefit from it.


Thanks,
Yi Chen

**Comment by aperloff** [comment]
I'm looking at it right now. Hopefully there aren't any more issues. I have to be a little careful because I/we don't have continuous integration set up. This means that I rely heavily on the people doing the PR to test the code to make sure there are no bugs.

File: JetAnalyzers/bin/compareJEC.C [hunk 2]
```diff
@@ -254,17 +254,21 @@ void compareJEC(string payld1="Winter14_V8", string payld2="",         string pa
...
   int maxTries = 7;
   string strPath;
-  vector<string> paths = {"CondFormats/JetMETObjects/data/"};
-  paths.push_back(string("/fdata/hepx/store/user/aperloff/JEC/80X_Summer16/")+cid1+"/");
-  paths.push_back(string("/fdata/hepx/store/user/aperloff/JEC/80X_Summer16/")+cid2+"/");
-  paths.push_back(string("/fdata/hepx/store/user/aperloff/JEC/80X_Summer16/")+cid3+"/");
-  paths.push_back(string("/home/aperloff/JEC/CMSSW_8_0_20/src/JetMETCorrections/JECDatabase/textFiles/")+cid1+"/");
-  paths.push_back(string("/home/aperloff/JEC/CMSSW_8_0_20/src/JetMETCorrections/JECDatabase/textFiles/")+cid2+"/");
-  paths.push_back(string("/home/aperloff/JEC/CMSSW_8_0_20/src/JetMETCorrections/JECDatabase/textFiles/")+cid3+"/");
+  vector<string> paths = {"","CondFormats/JetMETObjects/data/"};
```
Comment (aperloff) [inline JetAnalyzers/bin/compareJEC.C]: It would be nice to get the person's username from either a system call or an environment variable rather than hard coding it. Also, shouldn't the paths be a command line or configuration option? I just don't like that this is hard coded. Yes, I know I didn't change it before. Be better than me.

File: JetAnalyzers/bin/compareJEC.C [hunk 3]
```diff
@@ -1101,16 +1105,15 @@ void compareJEC(string payld1="Winter14_V8", string payld2="",         string pa
...
+  compareJEC("Fall17_25nsV1", "Summer16_25nsV5", "bias2SelectionPow_25nsV1", "AK4PFchs", "AK4PFchs", "AK4PFchs", "MC","MC","MC",    false, true,  false);
+  compareJEC("Fall17_25nsV1", "Summer16_25nsV5", "bias2SelectionPow_25nsV1", "AK4PFchs", "AK4PFchs", "AK4PFchs", "MC","MC","MC",    true, false,  false);
 
-  compareJEC("Summer16_25nsV5", "Summer16_25nsV4", "Spring16_25nsV6", "AK4PFPuppi", "AK4PFPuppi", "AK4PFPuppi", "MC","MC","MC",    false, true,  false);
-  compareJEC("Summer16_25nsV5", "Summer16_25nsV4", "Spring16_25nsV6", "AK4PFPuppi", "AK4PFPuppi", "AK4PFPuppi", "MC","MC","MC",    true, false,  false);
-
-  compareJEC("Summer16_25nsV5", "Summer16_25nsV4", "Spring16_25nsV6", "AK8PFPuppi", "AK8PFPuppi", "AK8PFPuppi", "MC","MC","MC",    false, true,  false);
-  compareJEC("Summer16_25nsV5", "Summer16_25nsV4", "Spring16_25nsV6", "AK8PFPuppi", "AK8PFPuppi", "AK8PFPuppi", "MC","MC","MC",    true, false,  false);
+  // compareJEC("Summer16_25nsV5", "Summer16_25nsV4", "Spring16_25nsV6", "AK8PFchs", "AK8PFchs", "AK8PFchs", "MC","MC","MC",    false, true,  false);
```
Comment (aperloff) [inline JetAnalyzers/bin/compareJEC.C]: Take out these commented lines if you're no longer using them.

File: JetAnalyzers/bin/jet_synchtest_x.cc [hunk 4]
```diff
@@ -795,6 +796,7 @@ void MatchEventsAndJets::LoopOverEvents(bool verbose, bool reduceHistograms, str
    for (IT::const_iterator it = mapTreePU.begin(); it != mapTreePU.end(); ++it) {
 
       if (iftest && nevs >= maxEvts) return;
+      // if (nevs >= 10000000) return;
```
Comment (aperloff) [inline JetAnalyzers/bin/jet_synchtest_x.cc]: Remove this line

File: JetAnalyzers/bin/jet_synchtest_x.cc [hunk 5]
```diff
@@ -795,6 +796,7 @@ void MatchEventsAndJets::LoopOverEvents(bool verbose, bool reduceHistograms, str
...
 
       //if (nevs%10000==0) cout << "\t"<<nevs << endl;
```
Comment (aperloff) [inline JetAnalyzers/bin/jet_synchtest_x.cc]: Remove this line too.

File: JetUtilities/interface/ClosureMaker.hh [hunk 6]
```diff
@@ -83,27 +83,28 @@ public:
 	void 				checkResponse();
 	pair<double,double> determineCanvasRange(double xmin, double xmax);
 	void 				makeCanvases();
-	void 				makeMergedCanvas();
+	void 				makeMergedCanvas(bool finemerge);
```
Comment (aperloff) [inline JetUtilities/interface/ClosureMaker.hh]: Any way to fix the spacing here? I know it's not your fault, but this would be a good opportunity to standardize the use of spaces and tabs.

File: JetUtilities/interface/HistogramUtilities.hh [hunk 7]
```diff
@@ -91,6 +92,10 @@ namespace HistUtil{
     //      (i.e. if the AxisDirection=Y, then the metric will be plotted along X)
     vector<pair<double,double> > getHistogramMetric2D(HistogramMetric, TH2*, AxisDirection, const vector<string>&);
     vector<pair<double,double> > getHistogramMetric2D(string, TH2*, AxisDirection, const vector<string>&);
+
+    void adjust_fitrange(TH1* h,double& min,double& max);
+    int number_filled_bins(TH1* h, double min, double max);
+    TF1* fit_gaussian(TH1*& hrsp, const double nsigma, const double jtptmin, const int niter, const int verbose);
```
Comment (aperloff) [inline JetUtilities/interface/HistogramUtilities.hh:98]: If you're going to add fit_gaussian here why not also add fit_dscb?

File: JetUtilities/interface/L2Creator.hh [hunk 8]
```diff
@@ -99,7 +99,7 @@ private:
     TString                   output, outputDir, l2calofit, l2pffit;
     vector<string>            formats, algs;
     bool                      l2l3, delphes;
-    int                       maxFitIter;
+    int                       maxFitIter, statTh;
```
Comment (aperloff) [inline JetUtilities/interface/L2Creator.hh]: It wasn't immediately clear to me from the name what this variable was. Can you call it "statThreshold"?

File: JetUtilities/src/ClosureMaker.cc [hunk 9]
```diff
@@ -385,10 +386,10 @@ void ClosureMaker::loopOverBins(TH2F* hvar, unsigned int iVarBin) {
                 continue;
             }
         }
-        else if(h.back()->GetEntries()<=4 && h.back()->GetEntries()>1) {
-            hClosure.back()->SetBinContent(ibin+1,h.back()->GetMean());
-            hClosure.back()->SetBinError(ibin+1,h.back()->GetMeanError());
-        }
+        // else if(h.back()->GetEntries()<=4 && h.back()->GetEntries()>1) {
```
Comment (aperloff) [inline JetUtilities/src/ClosureMaker.cc:389]: Why remove this?

File: JetUtilities/src/ClosureMaker.cc [hunk 10]
```diff
@@ -533,10 +534,10 @@ void ClosureMaker::makeCanvases() {
 }
 
 //______________________________________________________________________________
-void ClosureMaker::makeMergedCanvas() {
+void ClosureMaker::makeMergedCanvas(bool finemerge = false) {
```
Comment (aperloff) [inline JetUtilities/src/ClosureMaker.cc:537]: I don't love this finemerge variable. To me it just seems like a way to force the canvas range to be narrower. Can't you just change the settings in the function "determineCanvasRange"? Make that function smarter rather than coming up with a way to circumvent it.

File: JetUtilities/src/L2Creator.cc [hunk 11]
```diff
@@ -1373,16 +1378,17 @@ void L2Creator::writeTextFileForCurrentAlgorithm_spline() {
 
                 //For expediency of Summer16_25nsV5_MC do eta-dependent clipping
                 fout<<setw(8) <<etamin<<setw(8)<<etamax
-                    <<setw(10)<<setprecision(6)<<(isection ? bounds.first : 0.001)
+                    <<setw(10)<<setprecision(6)<<(firstline ? 0.001 : bounds.first)
```
Comment (aperloff) [inline JetUtilities/src/L2Creator.cc:1396]: Did you test this? Just want to make sure the logic is right.

File: JetUtilities/src/SynchFittingProcedure.hh [hunk 12]
```diff
@@ -345,34 +346,34 @@ TH1 * getMeanHistoFromHisto(TString cname, TString title, TH2 *off_in, double &
...
-    
+
 
       TH1 * aux= off_in->ProjectionY("_py",nb,nb);
       if (aux->GetEntries() > 0) {
 
-         TFitResultPtr fr = aux->Fit("gaus","0qS");
+         // TFitResultPtr fr = aux->Fit("gaus","0qS");
          //cout << cname << "sfsg1\tnb=" << nb << endl;
+         TF1* fr = HistUtil::fit_gaussian(aux, 1.5, 1.0, 10, false);
```
Comment (aperloff) [inline JetUtilities/src/SynchFittingProcedure.hh:364]: I love this. Thank you!

File: JetUtilities/src/ClosureMaker.cc (hunk 9 above)
Comment (sifuluo) [reply JetUtilities/src/ClosureMaker.cc:389]: There was a point in the closure whose response plot had exactly 4 entries, and it was not removed by statThreshold option. So I removed this line

File: JetUtilities/src/ClosureMaker.cc (hunk 10 above)
Comment (sifuluo) [reply JetUtilities/src/ClosureMaker.cc:537]: You are right, it is forcing the canvas to be narrowed to a range that high pT range should locate in. And also this is the plot people always want to see to make sure high pT region is fine. Same canvas range also makes it easier to compare between iterations.
There is such a determineCanvasRange function if I did not remember wrong. But when the closure it not good, it will try to incorporate all points in the range. making the range really large. This is of course useful to see the entire picture. So the finemerge is just in addition to this.

File: JetUtilities/src/ClosureMaker.cc (hunk 9 above)
Comment (aperloff) [reply JetUtilities/src/ClosureMaker.cc:389]: Well yeah, that's the logic set up there. What do you want to have happen?
1. If there are more than statThreshold entries and things proceed as normal
2. The part you commented out. If there is less than statThreshold entries, but it's not empty, do you want some fallback behavior? That was what these lines were for. If you don't want that behavior don't just comment these lines out; remove them entirely.
3. else skip this point

File: JetUtilities/src/ClosureMaker.cc (hunk 10 above)
Comment (aperloff) [reply JetUtilities/src/ClosureMaker.cc:537]: Why not make it so that if the dynamic range of the plot is too large it duplicates the plot and makes a "full range" version and a "zoomed in" version? Might be hard, but would make more sense from an ease of use and automation standpoint.

File: JetAnalyzers/bin/compareJEC.C (hunk 2 above)
Comment (sifuluo) [reply JetAnalyzers/bin/compareJEC.C]: I don't like it hard coded either, but files are not necessarily stored in /fdata/hepx/store/user/<username>/JEC/<ERA>/ . I store different JECs are different places and they all have their own patterns of paths. So it cannot be unified, at least not on my workflow. But I made changes so that it is not just for me.

File: JetUtilities/src/L2Creator.cc (hunk 11 above)
Comment (sifuluo) [reply JetUtilities/src/L2Creator.cc:1396]: Yes, the new corrections are all done by this

File: JetUtilities/src/ClosureMaker.cc (hunk 10 above)
Comment (sifuluo) [reply JetUtilities/src/ClosureMaker.cc:537]: It is hard to define 'too large'. Actually I think it would be best that we always have such 1.05-0.95 range canvas. So the too large would be defined as 'xmin<0.95 || xmax > 1.05'. And these days, it is almost always the case.

File: JetAnalyzers/bin/compareJEC.C (hunk 2 above)
Comment (aperloff) [reply JetAnalyzers/bin/compareJEC.C]: 1. Remove the commented out lines right below this. If they aren't needed then don't leave them lying around the code.
2. You can follow this SO post to replace your username with the one found by the system. https://stackoverflow.com/questions/8953424/how-to-get-the-username-in-c-c-in-linux That way it will work for whomever is using the code. Also, you can have a command line option to override the search locations, but only if the user specifies something on the command line.

File: JetUtilities/src/ClosureMaker.cc (hunk 10 above)
Comment (aperloff) [reply JetUtilities/src/ClosureMaker.cc:537]: Okay, then change the code to do that.

File: JetUtilities/src/L2Creator.cc (hunk 11 above)
Comment (aperloff) [reply JetUtilities/src/L2Creator.cc:1396]: Okay, then move onto the other comments.

**Comment by sifuluo** [comment]
The commented changed are made and these pieces of code is being used in Fall17 that proved its validity.

**Comment by aperloff** [comment]
@sifuluo There are plenty of comments that you haven't addressed. There are several about removing commented out lines. There's the comment about 'statThreshold'. There's the comment about getting the username from the environment variables. Please have a look. As soon as the comments are addressed I will merge the PR.

File: JetAnalyzers/bin/compareJEC.C (hunk 2 above)
Comment (sifuluo) [reply JetAnalyzers/bin/compareJEC.C]: I don't think it a good idea to have any preset convention of JEC paths, I sometimes compare my JECs with Andrew's, fixing username in a path does not seem to be useful for myself.

File: JetUtilities/interface/ClosureMaker.hh (hunk 6 above)
Comment (sifuluo) [reply JetUtilities/interface/ClosureMaker.hh]: I am converting all tabs to spaces when I modify a file now.

File: JetAnalyzers/bin/compareJEC.C (hunk 3 above)
Comment (sifuluo) [reply JetAnalyzers/bin/compareJEC.C]: done

File: JetAnalyzers/bin/jet_synchtest_x.cc (hunk 4 above)
Comment (sifuluo) [reply JetAnalyzers/bin/jet_synchtest_x.cc]: done

File: JetAnalyzers/bin/jet_synchtest_x.cc (hunk 5 above)
Comment (sifuluo) [reply JetAnalyzers/bin/jet_synchtest_x.cc]: done

File: JetUtilities/interface/HistogramUtilities.hh (hunk 7 above)
Comment (sifuluo) [reply JetUtilities/interface/HistogramUtilities.hh:98]: Because I am not sure what that is. And I have never got a chance to use this fit_dscb() yet.

File: JetUtilities/interface/L2Creator.hh (hunk 8 above)
Comment (sifuluo) [reply JetUtilities/interface/L2Creator.hh]: done

File: JetUtilities/src/ClosureMaker.cc (hunk 9 above)
Comment (sifuluo) [reply JetUtilities/src/ClosureMaker.cc:389]: If it does not exceed the statThreshold, I think it'd be good to skip the point to show that the correction is performing well. So I'll skip this point

**Comment by sifuluo** [comment]
I have replied to your comments, please check them out and inform me if there's anything else needed to be addressed.

File: JetAnalyzers/bin/compareJEC.C (hunk 2 above)
Comment (aperloff) [reply JetAnalyzers/bin/compareJEC.C]: Right now you have your username (and mine) hard coded in the the code. Either remove those lines completely or make it configurable. Also, the disk path (/fdata/hepx/) doesn't need to be hard coded. You can make it a command line option with a default. There are so many better ways of doing this than carrying along these hard coded paths.

File: JetUtilities/interface/HistogramUtilities.hh (hunk 7 above)
Comment (aperloff) [reply JetUtilities/interface/HistogramUtilities.hh:98]: It's a fit for a double sided crystal ball function. It performs the same type of operation as the fit_gaussian function, but for a different formula.

**Comment by aperloff** [comment]
@sifuluo Could this PR be split into multiple atomic PRs? I'd like to finish this so that I can get it off my plate, but at this point the PR is really hard to follow and covers a huge amount of code.
